components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+    /// let set = UnicodeSet::from_inversion_list(vec![0, 65, 70, (std::char::MAX as u32) + 1]).unwrap();
+    /// builder.add_set(&set);
+    /// builder.complement();
+    /// let check = builder.build();
+    /// assert_eq!(check.iter().next(), Some('A'));
+    /// ```
+    pub fn complement(&mut self) {
+        if !self.intervals.is_empty() {
+            if self.intervals[0] == 0 {
+                self.intervals.drain(0..1);
+            } else {
+                self.intervals.insert(0, 0);
+            }
+            if self.intervals.last() == Some(&((char::MAX as u32) + 1)) {
+                self.intervals.pop();
+            } else {
+                self.intervals.push((char::MAX as u32) + 1);
+            }
+        }
+    }
+
+    /// Complements the character in the builder, adding it if not in the builder, and removing it otherwise.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSetBuilder;
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// builder.add_range(&('A'..'D'));
+    /// builder.complement_char('A');
+    /// builder.complement_char('E');
+    /// let check = builder.build();
+    /// assert!(check.contains('E'));
+    /// assert!(!check.contains('A'));
+    /// ```
+    pub fn complement_char(&mut self, c: char) {
+        let code_point = c as u32;
+        let to_complement = vec![code_point, code_point + 1];
+        self.complement_list(to_complement);
+    }
+
+    /// Complements the range in the builder, adding any elements in the range if not in the builder, and
+    /// removing them otherwise.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSetBuilder;
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// builder.add_range(&('A'..'D'));
+    /// builder.complement_range(&('C'..='F'));
+    /// let check = builder.build();
+    /// assert!(check.contains('F'));
+    /// assert!(!check.contains('C'));
+    /// ```
+    pub fn complement_range(&mut self, range: &impl RangeBounds<char>) {
+        let (start, end) = deconstruct_range(range);
+        let to_complement = vec![start, end];
+        self.complement_list(to_complement);
+    }
+
+    /// Complements the set in the builder, adding any elements in the set if not in the builder, and
+    /// removing them otherwise.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::{UnicodeSetBuilder, UnicodeSet};
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// let set = UnicodeSet::from_inversion_list(vec![65, 70, 75, 90]).unwrap();
+    /// builder.add_range(&('C'..'N')); // 67 - 78
+    /// builder.complement_set(&set);
+    /// let check = builder.build();
+    /// assert!(check.contains('Q')); // 81
+    /// assert!(!check.contains('M')); // 77
+    /// ```
+    pub fn complement_set(&mut self, set: &UnicodeSet) {
+        self.complement_list(set.ranges().collect::<Vec<&[u32]>>().concat());
```

Comment:
Here, it would be nice if you could get the actual inversion list as a slice, so that you don't have to copy it into a new Vec.  You could consider adding a function to UnicodeSet that returns the inversion list slice, or you could make the field `pub(crate)`.

Reply:
`pub(crate)` was what I was looking for all this time... thank you, agreed and changed